### PR TITLE
List<AnyOf<*>> support

### LINKED ISSUE #25: Support List<AnyOf<Type>> fields with discriminator
👋

I've added a very limited support for `anyOf` when used in lists and combined with a discriminator.

More info on the topic:
https://swagger.io/docs/specification/data-models/inheritance-and-polymorphism/

## Problem

Here's an example schema file covering the issue:
https://github.com/vishna/example_openapi_anyof/blob/main/anyOf.yaml

The current implementation will generate a following `FruitBasket` class with this `fruits` field:

```dart
List<AnyOfmyApplemyBanana> fruits = [];
```

The generated package doesn't compile since the class `AnyOfmyApplemyBanana` doesn't exists.

##  Solution

The fix generates code for the missing class, it looks something like this:

```dart
class AnyOfMyAppleMyBanana {
  AnyOfMyAppleMyBanana();

  dynamic _value;
  dynamic get value => _value;

  @override
  String toString() {
    return 'AnyOf<MyApple,MyBanana>[value=$_value]';
  }

  fromJson(Map<String, dynamic> json) {
    if (json == null) return;

    final discriminator = json[r'type'];

    if (discriminator == null) {
      return;
    }

    switch (discriminator) {
      case r'apple':
        _value = MyApple.fromJson(json);
        break;
      case r'banana':
        _value = MyBanana.fromJson(json);
        break;
    }
  }

  AnyOfMyAppleMyBanana.fromJson(Map<String, dynamic> json) {
    fromJson(json); // allows child classes to call
  }

  Map<String, dynamic> toJson() => LocalApiClient.serialize(_value);

  @override
  bool operator ==(Object other) {
    if (identical(this, other)) {
      return true;
    }

    if (other is AnyOfMyAppleMyBanana && runtimeType == other.runtimeType) {
      return _value == other._value;
    }

    return false;
  }

  @override
  int get hashCode {
    var hashCode = runtimeType.hashCode;

    if (_value != null) {
      hashCode = hashCode * 31 + _value.hashCode;
    }

    return hashCode;
  }

  AnyOfMyAppleMyBanana copyWith() {
    final copy = AnyOfMyAppleMyBanana();
    copy._value = _value?.copyWith();
    return copy;
  }

  static List<AnyOfMyAppleMyBanana> listFromJson(List<dynamic> json) {
    return json == null
        ? <AnyOfMyAppleMyBanana>[]
        : json.map((value) => AnyOfMyAppleMyBanana.fromJson(value)).toList();
  }
}
```

I've added a sample project with a building script here:
https://github.com/vishna/example_openapi_anyof

### PR DESCRIPTION
**Pull request by vishna**
Resolves #25 

I have no deep knowledge of this codebase and this PR is pretty much a result of poking around different parts of the ApiGenerator.

Essentially my hacking starts with a method called `toAnyOfName`. This method seems to be triggered for every __List<AnyOf>__ and before mustache files are rendered. This is where I intercept __anyOf__ occurrences and store this data for later in `extraAnyParts` and `extraAnyClasses` fields. Then I use these fields to populate mustache files.

I got something working so I'm happy as it solves my issue. Unsure though if maintainers are happy with the quality of my code 😅

# Checklist

- [x] Add enum to class and include in the hash - should be also be useful for this:

```
switch(myValue.discriminator) {
  case AnyOfAB.A:
  case AnyOfAB.B:
}
```
- [x] Added value setter (useful when writing tests that involve AnyOf generated classes)
- [x] Added list_any_of.yaml test scenario to `SampleRunner`
- [x] keeping _value as non-nullable, if the value is supposed to be nullable, it should be `AnyOfClass?`
- [x] integration tests (verify package from `SampleRunner`)
- [x] Check if our internal generated API package doesn't get broken ( 🍯 https://github.com/vishna/dart-openapi-maven/releases/tag/openapi-dart-generator-5.3-anyOf.1 )

### CODE REVIEW

**Comment by rvowles** [comment]
Code quality is perfectly nice :-) Just need to add some examples to the it/* integration test suites and make sure they compile and I can merge it and release it :-)

**Comment by vishna** [comment]
I will finally be working on this tomorrow 🤞

**Comment by rvowles** [comment]
There is a new Github Action that should test this as you push

**Comment by vishna** [comment]
@rvowles so this should be mostly ready now, I've also added a simple `verifyPackage` function to the `SampleRunner` - checks correctness from dart point of view after generation. Shouldn't `SampleRunner` also be a part of github action?

**Comment by vishna** [comment]
...ah and need workflow approval, otherwise the action doesn't ever get triggered.

**Comment by rvowles** [comment]
Sample runner is only for development purposes, it's not used for testing on release. That's what the example projects are for in the it (integration testing) folder, they have actual dart tests in them that get run as part of the build that you should add tests to.

**Comment by vishna** [comment]
Thanks, so that's what I also did. Now just waiting for this to turn green 💚
Screenshot 2021-05-11 at 08 47 32

File: README.adoc [hunk 1]
```diff
@@ -305,3 +305,13 @@ behind as you generate.
       </plugin>
 ----
 
+=== Testing
+
+To test locally you can run tests by invoking this command:
+
+[source,bash]
+----
+mvn -Dmaven.wagon.http.retryHandler.count=5 --batch-mode verify
```
Comment (rvowles) [inline README.adoc:314]: this should be `mvn clean verify` - the other stuff is if you have a clean build and unreliable build process like GitHub is :-(

**Pull request merged**